src/sinks/loki.rs

```diff
@@ -0,0 +1,446 @@
+//! Loki sink
+//!
+//! This sink provides downstream support for `Loki` via
+//! the v1 http json endpoint.
+//!
+//! https://github.com/grafana/loki/blob/master/docs/api.md
+//!
+//! This sink does not use `PartitionBatching` but elects to do
+//! stream multiplexing by organizing the streams in the `build_request`
+//! phase. There must be at least one valid set of labels.
+//!
+//! If an event produces no labels, this can happen if the template
+//! does not match, we will add a default label `{agent="vector"}`.
+
+use crate::{
+    dns::Resolver,
+    event::{self, Event, Value},
+    sinks::util::http::{https_client, Auth, BatchedHttpSink, HttpSink},
+    sinks::util::tls::{TlsOptions, TlsSettings},
+    sinks::util::{BatchBytesConfig, TowerRequestConfig, UriSerde},
+    template::Template,
+    topology::config::{DataType, SinkConfig, SinkContext, SinkDescription},
+};
+use derivative::Derivative;
+use futures03::{compat::Future01CompatExt, TryFutureExt};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::collections::HashMap;
+
+type Labels = Vec<(String, String)>;
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct LokiConfig {
+    host: UriSerde,
+    encoding: Encoding,
+
+    tenant_id: Option<String>,
+    labels: HashMap<String, Template>,
+
+    remove_labels: Option<bool>,
+    remove_timestamp: Option<bool>,
+
+    auth: Option<Auth>,
+
+    #[serde(default)]
+    request: TowerRequestConfig,
+
+    #[serde(default)]
+    batch: BatchBytesConfig,
+
+    tls: Option<TlsOptions>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, Derivative)]
+#[serde(rename_all = "snake_case")]
+#[derivative(Default)]
+enum Encoding {
+    #[derivative(Default)]
+    Json,
+    Text,
+}
+
+inventory::submit! {
+    SinkDescription::new_without_default::<LokiConfig>("loki")
+}
+
+#[typetag::serde(name = "loki")]
+impl SinkConfig for LokiConfig {
+    fn build(&self, cx: SinkContext) -> crate::Result<(super::RouterSink, super::Healthcheck)> {
+        if self.labels.is_empty() {
+            return Err(format!("`labels` must include atleast one label.").into());
+        }
+
+        let request_settings = self.request.unwrap_with(&TowerRequestConfig::default());
+        let batch_settings = self.batch.unwrap_or(bytesize::mib(10u64), 1);
+        let tls = TlsSettings::from_options(&self.tls)?;
+
+        let sink = BatchedHttpSink::new(
+            self.clone(),
+            Vec::new(),
+            request_settings,
+            batch_settings,
+            Some(tls),
+            &cx,
+        );
+
+        let healthcheck = Box::new(Box::pin(healthcheck(self.clone(), cx.resolver())).compat());
```

Comment:
Nitpick, but this might be a little easier to read if we did

```rust
let healthcheck = Box::pin(healthcheck(self.clone(), cx.resolver()));

Ok((Box::new(sink), Box::new(healthcheck.compat()))
```